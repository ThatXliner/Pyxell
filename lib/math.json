{
	"node": "Block",
	"position": [
		2,
		1
	],
	"stmts": [
		{
			"node": "StmtAssg",
			"position": [
				2,
				1
			],
			"lvalues": [
				{
					"node": "AtomId",
					"position": [
						2,
						1
					],
					"id": "PI"
				}
			],
			"expr": {
				"node": "AtomFloat",
				"position": [
					2,
					6
				],
				"float": 3.141592653589793
			}
		},
		{
			"node": "StmtAssg",
			"position": [
				3,
				1
			],
			"lvalues": [
				{
					"node": "AtomId",
					"position": [
						3,
						1
					],
					"id": "E"
				}
			],
			"expr": {
				"node": "AtomFloat",
				"position": [
					3,
					5
				],
				"float": 2.718281828459045
			}
		},
		{
			"node": "StmtAssg",
			"position": [
				4,
				1
			],
			"lvalues": [
				{
					"node": "AtomId",
					"position": [
						4,
						1
					],
					"id": "INF"
				}
			],
			"expr": {
				"node": "ExprCall",
				"position": [
					4,
					7
				],
				"expr": {
					"node": "ExprAttr",
					"position": [
						4,
						7
					],
					"expr": {
						"node": "AtomString",
						"position": [
							4,
							7
						],
						"string": "inf"
					},
					"attr": "toFloat",
					"safe": false
				},
				"args": []
			}
		},
		{
			"node": "StmtFunc",
			"position": [
				6,
				1
			],
			"id": "exp",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						6,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							6,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					6,
					19
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				7,
				1
			],
			"id": "log",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						7,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							7,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					7,
					19
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				8,
				1
			],
			"id": "log10",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						8,
						12
					],
					"type": {
						"node": "TypeName",
						"position": [
							8,
							12
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					8,
					21
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				9,
				1
			],
			"id": "sqrt",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						9,
						11
					],
					"type": {
						"node": "TypeName",
						"position": [
							9,
							11
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					9,
					20
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				11,
				1
			],
			"id": "cos",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						11,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							11,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					11,
					19
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				12,
				1
			],
			"id": "sin",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						12,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							12,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					12,
					19
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				13,
				1
			],
			"id": "tan",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						13,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							13,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					13,
					19
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				15,
				1
			],
			"id": "acos",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						15,
						11
					],
					"type": {
						"node": "TypeName",
						"position": [
							15,
							11
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					15,
					20
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				16,
				1
			],
			"id": "asin",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						16,
						11
					],
					"type": {
						"node": "TypeName",
						"position": [
							16,
							11
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					16,
					20
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				17,
				1
			],
			"id": "atan",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						17,
						11
					],
					"type": {
						"node": "TypeName",
						"position": [
							17,
							11
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					17,
					20
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				18,
				1
			],
			"id": "atan2",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						18,
						12
					],
					"type": {
						"node": "TypeName",
						"position": [
							18,
							12
						],
						"name": "Float"
					},
					"name": "y",
					"default": null
				},
				{
					"node": "FuncArg",
					"position": [
						18,
						21
					],
					"type": {
						"node": "TypeName",
						"position": [
							18,
							21
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					18,
					30
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				20,
				1
			],
			"id": "rad",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						20,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							20,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					20,
					19
				],
				"name": "Float"
			},
			"block": {
				"node": "Block",
				"position": [
					20,
					28
				],
				"stmts": [
					{
						"node": "StmtReturn",
						"position": [
							21,
							5
						],
						"expr": {
							"node": "ExprBinaryOp",
							"position": [
								21,
								12
							],
							"exprs": [
								{
									"node": "ExprBinaryOp",
									"position": [
										21,
										12
									],
									"exprs": [
										{
											"node": "AtomId",
											"position": [
												21,
												12
											],
											"id": "x"
										},
										{
											"node": "AtomId",
											"position": [
												21,
												16
											],
											"id": "PI"
										}
									],
									"op": "*"
								},
								{
									"node": "AtomInt",
									"position": [
										21,
										21
									],
									"int": 180
								}
							],
							"op": "/"
						}
					}
				]
			}
		},
		{
			"node": "StmtFunc",
			"position": [
				23,
				1
			],
			"id": "deg",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						23,
						10
					],
					"type": {
						"node": "TypeName",
						"position": [
							23,
							10
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					23,
					19
				],
				"name": "Float"
			},
			"block": {
				"node": "Block",
				"position": [
					23,
					28
				],
				"stmts": [
					{
						"node": "StmtReturn",
						"position": [
							24,
							5
						],
						"expr": {
							"node": "ExprBinaryOp",
							"position": [
								24,
								12
							],
							"exprs": [
								{
									"node": "ExprBinaryOp",
									"position": [
										24,
										12
									],
									"exprs": [
										{
											"node": "AtomId",
											"position": [
												24,
												12
											],
											"id": "x"
										},
										{
											"node": "AtomInt",
											"position": [
												24,
												16
											],
											"int": 180
										}
									],
									"op": "*"
								},
								{
									"node": "AtomId",
									"position": [
										24,
										22
									],
									"id": "PI"
								}
							],
							"op": "/"
						}
					}
				]
			}
		},
		{
			"node": "StmtFunc",
			"position": [
				26,
				1
			],
			"id": "floor",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						26,
						12
					],
					"type": {
						"node": "TypeName",
						"position": [
							26,
							12
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					26,
					21
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				27,
				1
			],
			"id": "ceil",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						27,
						11
					],
					"type": {
						"node": "TypeName",
						"position": [
							27,
							11
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					27,
					20
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				28,
				1
			],
			"id": "trunc",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						28,
						12
					],
					"type": {
						"node": "TypeName",
						"position": [
							28,
							12
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					28,
					21
				],
				"name": "Float"
			},
			"block": null
		},
		{
			"node": "StmtFunc",
			"position": [
				30,
				1
			],
			"id": "round",
			"typevars": [],
			"args": [
				{
					"node": "FuncArg",
					"position": [
						30,
						12
					],
					"type": {
						"node": "TypeName",
						"position": [
							30,
							12
						],
						"name": "Float"
					},
					"name": "x",
					"default": null
				},
				{
					"node": "FuncArg",
					"position": [
						30,
						21
					],
					"type": {
						"node": "TypeName",
						"position": [
							30,
							21
						],
						"name": "Int"
					},
					"name": "p",
					"default": {
						"node": "AtomInt",
						"position": [
							30,
							28
						],
						"int": 0
					}
				}
			],
			"ret": {
				"node": "TypeName",
				"position": [
					30,
					31
				],
				"name": "Float"
			},
			"block": {
				"node": "Block",
				"position": [
					30,
					40
				],
				"stmts": [
					{
						"node": "StmtIf",
						"position": [
							31,
							5
						],
						"exprs": [
							{
								"node": "ExprCmp",
								"position": [
									31,
									8
								],
								"exprs": [
									{
										"node": "AtomId",
										"position": [
											31,
											8
										],
										"id": "p"
									},
									{
										"node": "AtomInt",
										"position": [
											31,
											13
										],
										"int": 0
									}
								],
								"ops": [
									"!="
								]
							}
						],
						"blocks": [
							{
								"node": "Block",
								"position": [
									31,
									17
								],
								"stmts": [
									{
										"node": "StmtAssgExpr",
										"position": [
											32,
											9
										],
										"exprs": [
											{
												"node": "AtomId",
												"position": [
													32,
													9
												],
												"id": "x"
											},
											{
												"node": "ExprBinaryOp",
												"position": [
													32,
													14
												],
												"exprs": [
													{
														"node": "AtomFloat",
														"position": [
															32,
															14
														],
														"float": 10.0
													},
													{
														"node": "AtomId",
														"position": [
															32,
															19
														],
														"id": "p"
													}
												],
												"op": "^"
											}
										],
										"op": "*"
									}
								]
							}
						]
					},
					{
						"node": "StmtAssg",
						"position": [
							33,
							5
						],
						"lvalues": [
							{
								"node": "AtomId",
								"position": [
									33,
									5
								],
								"id": "x"
							}
						],
						"expr": {
							"node": "ExprCond",
							"position": [
								33,
								9
							],
							"exprs": [
								{
									"node": "ExprCmp",
									"position": [
										33,
										9
									],
									"exprs": [
										{
											"node": "AtomId",
											"position": [
												33,
												9
											],
											"id": "x"
										},
										{
											"node": "AtomFloat",
											"position": [
												33,
												13
											],
											"float": 0.0
										}
									],
									"ops": [
										">"
									]
								},
								{
									"node": "ExprCall",
									"position": [
										33,
										19
									],
									"expr": {
										"node": "AtomId",
										"position": [
											33,
											19
										],
										"id": "floor"
									},
									"args": [
										{
											"node": "CallArg",
											"position": [
												33,
												25
											],
											"name": null,
											"expr": {
												"node": "ExprBinaryOp",
												"position": [
													33,
													25
												],
												"exprs": [
													{
														"node": "AtomId",
														"position": [
															33,
															25
														],
														"id": "x"
													},
													{
														"node": "AtomFloat",
														"position": [
															33,
															27
														],
														"float": 0.5
													}
												],
												"op": "+"
											}
										}
									]
								},
								{
									"node": "ExprCall",
									"position": [
										33,
										34
									],
									"expr": {
										"node": "AtomId",
										"position": [
											33,
											34
										],
										"id": "ceil"
									},
									"args": [
										{
											"node": "CallArg",
											"position": [
												33,
												39
											],
											"name": null,
											"expr": {
												"node": "ExprBinaryOp",
												"position": [
													33,
													39
												],
												"exprs": [
													{
														"node": "AtomId",
														"position": [
															33,
															39
														],
														"id": "x"
													},
													{
														"node": "AtomFloat",
														"position": [
															33,
															41
														],
														"float": 0.5
													}
												],
												"op": "-"
											}
										}
									]
								}
							]
						}
					},
					{
						"node": "StmtIf",
						"position": [
							34,
							5
						],
						"exprs": [
							{
								"node": "ExprCmp",
								"position": [
									34,
									8
								],
								"exprs": [
									{
										"node": "AtomId",
										"position": [
											34,
											8
										],
										"id": "p"
									},
									{
										"node": "AtomInt",
										"position": [
											34,
											13
										],
										"int": 0
									}
								],
								"ops": [
									"!="
								]
							}
						],
						"blocks": [
							{
								"node": "Block",
								"position": [
									34,
									17
								],
								"stmts": [
									{
										"node": "StmtAssgExpr",
										"position": [
											35,
											9
										],
										"exprs": [
											{
												"node": "AtomId",
												"position": [
													35,
													9
												],
												"id": "x"
											},
											{
												"node": "ExprBinaryOp",
												"position": [
													35,
													14
												],
												"exprs": [
													{
														"node": "AtomFloat",
														"position": [
															35,
															14
														],
														"float": 10.0
													},
													{
														"node": "AtomId",
														"position": [
															35,
															19
														],
														"id": "p"
													}
												],
												"op": "^"
											}
										],
										"op": "/"
									}
								]
							}
						]
					},
					{
						"node": "StmtReturn",
						"position": [
							36,
							5
						],
						"expr": {
							"node": "AtomId",
							"position": [
								36,
								12
							],
							"id": "x"
						}
					}
				]
			}
		}
	]
}