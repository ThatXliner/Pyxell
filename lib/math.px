
PI = 3.1415926535897932384626
E = 2.7182818284590452353603
INF = "inf".toFloat()

func exp(Float x) Float extern
func log(Float x) Float extern
func log10(Float x) Float extern
func sqrt(Float x) Float extern

func cos(Float x) Float extern
func sin(Float x) Float extern
func tan(Float x) Float extern

func acos(Float x) Float extern
func asin(Float x) Float extern
func atan(Float x) Float extern
func atan2(Float y, Float x) Float extern

func rad(Float x) Float def
    return x * PI / 180

func deg(Float x) Float def
    return x * 180 / PI

func floor(Float x) Float extern
func ceil(Float x) Float extern
func trunc(Float x) Float extern

func round(Float x, Int p: 0) Float def
    Float m = 10^^abs(p)
    if p < 0 do
        m = 1.0 / m
    if p != 0 do
        x *= m
    x = x > 0.0 ? floor(x+0.5) : ceil(x-0.5)
    if p != 0 do
        x /= m
    return x
