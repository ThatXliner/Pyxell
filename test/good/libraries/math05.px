
use math

print round(4.2), round(987.5), round(-46.18), round(-1734.5)
print round(1234.5678, 1), round(-0.8273449, 4), round(0.0007, 2)
print round(-43.5, -1), round(538, -3)
