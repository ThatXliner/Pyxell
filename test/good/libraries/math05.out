4.0 988.0 -46.0 -1735.0
1234.6 -0.8273 0.0
-40.0 1000.0
